### DISCUSSION TITLE
Is there a method to base a linked resource on an environment variable?

### DISCUSSION THREAD

**Discussion by gvl2023** [eclipse-platform/eclipse.platform#340] [Q&A]
I know that linked resources can be created using internal Eclipse variables. Is it somehow possible to create a linked resource using a path that's contained in an environment variable?

**Comment by jukzi** [comment]
no - environment variables can change at runtime. you can use only theese constants:
![image](https://user-images.githubusercontent.com/51790620/221773109-58f27881-c7d0-4a7c-8983-1fe1addb372e.png)

**Reply by laeubi** [reply]
> no - environment variables can change at runtime.

How?

Linux: https://unix.stackexchange.com/questions/38205/change-environment-of-a-running-process
Windows: https://superuser.com/questions/1141461/how-to-change-environment-variable-to-running-process-in-windows

So I think env variables might change over time, but once the process is started they can't change unless you restart the process, so they are as "static" as the ECLIPSE_HOME (that possibly can change as well).

**Reply by gvl2023** [reply]
Yeah, I don't understand this argument. Even if they can be changed for a running process this is not how they are used.

**Reply by merks** [reply]
The point too to consider how this can be portable to different machines and OSes when you check it into git and check it out elsewhere.  Maybe that’s also not how you use it.  But it is a general consideration.

**Reply by jukzi** [reply]
Dynamically Add/Edit Environment Variables of Remote Process
https://www.codeproject.com/Articles/32131/Dynamically-Add-Edit-Environment-Variables-of-Remo

**Reply by laeubi** [reply]
Well this is really nothing that qualifies ENV variables as "they can change dynamically"... I assume that there is actually no process out there that really assumes they change "dynamically"....

**Reply by gvl2023** [reply]
Seems like a wrong reason to omit a very useful feature if this is why there’s no support for env. variables out of the box.

**Reply by mickaelistria** [reply]
I think support for environment variables in linked resources path wouldn't be rejected if you contribute it.

**Reply by iloveeclipse** [reply]
> Seems like a wrong reason to omit a very useful feature if this is why there’s no support for env. variables out of the box.

No one knows why that wasn't implemented at very beginning. May be the original authors of the code simply didn't assumed this as useful at all, because they were running on Windows and the environment variables were not so common in their own development. We don't know that anymore, because no one from original authors of the code is there.

But assuming there is a way to properly handle duplicated variables provided by different providers, one can add one default provider in IDE (may be guarded by an option) that would simply mirror all of the environment variables.

**Comment by iloveeclipse** [comment]
One can contribute custom path variable resolver, see https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fextension-points%2Forg_eclipse_core_resources_variableResolvers.html

This one can simply mirror all environment variables.

Note: since the variables returned by resolver can be saved & used in project files, one should supply only variables that are "known to be stable" in the current environment.

**Reply by mickaelistria** [reply]
Debug allows `${env_var:PATH}` or similar, Maven allows `env.PATH`... I would recommend someone who wants to contribute such variable resolver to use one of these prefix.

**Reply by laeubi** [reply]
You don't need to mirror all variables... a variable can be parameterized (see the "extend" button in the screenshot above) so one only needs ONE new ENV variable that takes the desired variable as a parameter.

**Reply by gvl2023** [reply]
> Debug allows `${env_var:PATH}` or similar, Maven allows `env.PATH`... I would recommend someone who wants to contribute such variable resolver to use one of these prefix.

Yup, I was going to suggest the same.

**Comment by merks** [comment]
One significant consideration is that if you checkin such a project with such links, how well does that work on another machine and/or even another OS?

**Reply by iloveeclipse** [reply]
This is a problem of people who **use** the variable, not of those who provide a variable, isn't?

**Reply by laeubi** [reply]
This is not really the concern of that function. Git itself support links, that only work on linux/mac but not on windows, different OS has different encodings, you can even at the moment parameterize the workspace loc with a random path in your workspace that must not be checked in anywhere, can be even DYNAMICALLY(!) deleted and so on ... there is so many things that can go wrong currently but still people are able to work together so I really don't think one should be over concerned here:
![grafik](https://user-images.githubusercontent.com/1331477/223158389-5ef7151d-9d6e-401c-90c3-6b0cdc067979.png)

**Reply by merks** [reply]
Providing something that’s at best problematic to use certainly is not a ringing endorsement to providing it and is perhaps the reason it hasn’t been provided.  I only mention this as something to consider, not as a reason to block this.
